compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala

```diff
@@ -0,0 +1,281 @@
+package dotty.tools.dotc
+package transform
+
+import java.io.File
+import java.util.concurrent.atomic.AtomicInteger
+
+import collection.mutable
+import core.Flags.*
+import core.Contexts.{Context, ctx, inContext}
+import core.DenotTransformers.IdentityDenotTransformer
+import core.Symbols.{defn, Symbol}
+import core.Decorators.{toTermName, i}
+import core.Constants.Constant
+import core.NameOps.isContextFunction
+import core.Types.*
+import typer.LiftCoverage
+import util.{SourcePosition, Property}
+import util.Spans.Span
+import coverage.*
+
+/** Implements code coverage by inserting calls to scala.runtime.Invoker
+  * ("instruments" the source code).
+  * The result can then be consumed by the Scoverage tool.
+  */
+class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
+  import ast.tpd._
+
+  override def phaseName = InstrumentCoverage.name
+
+  override def description = InstrumentCoverage.description
+
+  // Enabled by argument "-coverage OUTPUT_DIR"
+  override def isEnabled(using ctx: Context) =
+    ctx.settings.coverageOutputDir.value.nonEmpty
+
+  // Atomic counter used for assignation of IDs to difference statements
+  private val statementId = AtomicInteger(0)
+
+  private var outputPath = ""
+
+  // Main class used to store all instrumented statements
+  private val coverage = Coverage()
+
+  override def run(using ctx: Context): Unit =
+    outputPath = ctx.settings.coverageOutputDir.value
+
+    // Ensure the dir exists
+    val dataDir = new File(outputPath)
+    val newlyCreated = dataDir.mkdirs()
+
+    if !newlyCreated then
+      // If the directory existed before, let's clean it up.
+      dataDir.listFiles.nn
+        .filter(_.nn.getName.nn.startsWith("scoverage"))
+        .foreach(_.nn.delete())
+    end if
+    super.run
+
+    Serializer.serialize(coverage, outputPath, ctx.settings.coverageSourceroot.value)
+
+  override protected def newTransformer(using Context) = CoverageTransormer()
+
+  /** Transforms trees to insert calls to Invoker.invoked to compute the coverage when the code is called */
+  private class CoverageTransormer extends Transformer:
+    private val IgnoreLiterals = new Property.Key[Boolean]
+
+    private def ignoreLiteralsContext(using ctx: Context): Context =
+      ctx.fresh.setProperty(IgnoreLiterals, true)
+
+    override def transform(tree: Tree)(using ctx: Context): Tree =
+      tree match
+        // simple cases
+        case tree: (Import | Export | This | Super | New) => tree
+        case tree if (tree.isEmpty || tree.isType) => tree // empty Thicket, Ident, TypTree, ...
+
+        // Literals must be instrumented (at least) when returned by a def,
+        // otherwise `def d = "literal"` is not covered when called from a test.
+        // They can be left untouched when passed in a parameter of an Apply.
```

Comment:
Is this because the whole Apply block will be marked as instrumented if the method is called?